Выполнение ТЗ№10

EvgenEng opened the pull request.

EvgenEng committed:
Выполнение ТЗ№10

EvgenEng committed:
Remove checkstyle.xml as it's not allowed to be modified

EvgenEng committed:
исправление ошибки 254 из автотеста (удален из PR checkstyle.

EvgenEng committed:
Revert checkstyle.xml changes

EvgenEng committed:
Исправлены контроллеры и обработка ошибок: разделены ValidationException и NotFoundException, добавлен ErrorResponse

EvgenEng committed:
Исправлены контроллеры и обработка ошибок

afomkina commented on src/main/java/ru/yandex/practicum/controllers/ErrorHandler.java:
```diff
@@ -0,0 +1,60 @@
+public class ErrorHandler {
+
+    @ExceptionHandler(ValidationException.class)
+    public ResponseEntity<Map<String, Object>> handleValidationException(ValidationException e) {
```
Можно сделать код лаконичнее, избавившись от обертки `ResponseEntity`. Для этого нужно просто добавить аннотацию `@ResponseStatus(HttpStatus.NOT_FOUND)` над методом, а в самом методе. А вместо <Map<String, Object>> лучше создавть класс `ErrorResponse` с нужными полями

afomkina commented on src/main/java/ru/yandex/practicum/controllers/ErrorHandler.java:
```diff
@@ -0,0 +1,60 @@
+    }
+
+    @ExceptionHandler(NotFoundException.class)
+    public ResponseEntity<Map<String, Object>> handleNotFoundException(NotFoundException e) {
```
Нужно избавиться от однобуквенных названий переменных

afomkina commented on src/main/java/ru/yandex/practicum/controllers/FilmController.java:
```diff
@@ -0,0 +1,79 @@
+
+    private void validateFilm(Film film) {
+        // Валидация даты релиза
+        if (film.getReleaseDate().isBefore(LocalDate.of(1895, 12, 28))) {
```
Можно реализовать кастомную аннотацию валидации, чтобы полностью отказаться от ручной валидации.
Вот [отличная статья](https://reflectoring.io/bean-validation-with-spring-boot/#a-custom-validator-with-spring-boot:~:text=to%20none.-,A%20Custom%20Validator%20with%20Spring%20Boot,-If%20the%20available), в которой подробно описано, как это можно сделать

afomkina commented on src/main/java/ru/yandex/practicum/controllers/FilmController.java:
```diff
@@ -0,0 +1,79 @@
+        }
+
+        // Валидация описания
+        if (film.getDescription() != null && film.getDescription().length() > 200) {
```
Так как у тебя используются аннотации валидации, всю ручную валидацию нужно удалить из проекта. Здесь и во всем остальном коде валидация должна работать исключительно с помощью аннотаций

afomkina commented on src/main/java/ru/yandex/practicum/controllers/UserController.java:
```diff
@@ -0,0 +1,60 @@
+    @PostMapping
+    public User create(@Valid @RequestBody User user) {
+        // Автозаполнение имени, если оно пустое
+        if (user.getName() == null || user.getName().isBlank()) {
```
Нужно перенести эту проверку в конструктор или сеттер класса `User`

afomkina commented on src/main/java/ru/yandex/practicum/exception/ErrorResponse.java:
```diff
@@ -0,0 +1,15 @@
+
+@Data
+@AllArgsConstructor
+public class ErrorResponse {
```
Нужно использовать в обработчиках ошибок

afomkina commented on src/test/java/ru/yandex/practicum/controllers/UserControllerTest.java:
```diff
@@ -0,0 +1,102 @@
+    }
+
+    @Test
+    void shouldUseLoginAsNameWhenNameIsEmpty() {
```
Для тестирования спринговой валидации можно воспользоваться [валидатором](https://stackoverflow.com/questions/29069956/how-to-test-validation-annotations-of-a-class-using-junit)

EvgenEng committed:
Оптимизация обработки ошибок и валидации:
- Упрощены обработчики исключений с использованием класса ErrorResponse.
- Улучшена читаемость кода за счет переименования переменных.
- Удалена ручная валидация в FilmController, внедрены кастомные аннотации.
- Перенесена логика проверки имени пользователя в конструктор класса User.
- Добавлен валидатор для тестирования спринговой валидации.

afomkina commented on src/main/java/ru/yandex/practicum/validators/ReleaseDateValidator.java:
```diff
@@ -0,0 +1,14 @@
+import jakarta.validation.ConstraintValidatorContext;
+import java.time.LocalDate;
+
+public class ReleaseDateValidator implements ConstraintValidator<ValidReleaseDate, LocalDate> {
```
Получилось здорово, молодец 👍

afomkina commented on src/test/java/ru/yandex/practicum/controllers/FilmControllerTest.java:
```diff
@@ -0,0 +1,89 @@
+        Film created = restTemplate.postForEntity("/films", testFilm, Film.class).getBody();
+        created.setName("Updated Film");
+
+        ResponseEntity<Film> response = restTemplate.postForEntity("/films", created, Film.class);
```
Здесь и ниже тесты ничего не проверяют. Чтобы проверить, что была выброшена ошибка, нужно использовать [assertThrows](https://www.baeldung.com/junit-assert-exception)

afomkina commented on suppressions.xml:
Этот код нужно удалить

EvgenEng committed:
Удаление файла suppressions.xml

EvgenEng committed:
Исправлены тесты в соответствии с замечаниями

afomkina commented on suppressions.xml:
Извини, пожалуйста, я погорячилась. Полностью удалять этот файл не нужно. В нем просто не должно быть изменений, то есть код должен быть:
```
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE suppressions PUBLIC
		"-//Checkstyle//DTD SuppressionFilter Configuration 1.2//EN"
		"https://checkstyle.org/dtds/suppressions_1_2.dtd">
<suppressions>
	<suppress checks=".*" files="target[\\/]generated-sources[\\/]"/>
</suppressions>
```

afomkina commented on src/test/java/ru/yandex/practicum/controllers/UserControllerTest.java:
```diff
@@ -0,0 +1,112 @@
+    }
+
+    @Test
+    void shouldReturnNotFoundForUnknownUserUpdate() {
```
Название теста говорит, что ожидается ошибка, но приходит OK

EvgenEng replied:
скорректировал

EvgenEng committed:
Восстановлен suppressions.xml

EvgenEng committed:
Исправлен тест shouldReturnNotFoundForUnknownUserUpdate и восстановлен файл shouldReturnNotFoundForUnknownUserUpdate в РЕПО
